### PR TITLE
⚠️ Bastion cleanup

### PR DESCRIPTION
**Pull request by dulek**
**What this PR does / why we need it**:
This commit works with .spec.bastion:
* makes availablityZone, floatingIP and instance pointers
* adds floatingIP IPv4 validation
* adds validation that if Bastion is enabled, instance cannot be null
* rename instance->spec

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**TODOs**:

- if necessary:
  - [X] includes documentation
  - [X] adds unit tests

/hold

### CODE REVIEW

File: api/v1beta1/types.go [hunk 1]
```diff
@@ -596,21 +596,26 @@ var (
 	InstanceStateUndefined = InstanceState("")
 )
 
-// Bastion represents basic information about the bastion node.
+// Bastion represents basic information about the bastion node. If you enable bastion, the spec has to be specified.
+// +kubebuilder:validation:XValidation:rule="!self.enabled || has(self.spec)",message="you need to specify the spec if bastion is enabled"
 type Bastion struct {
+	// Enabled means that bastion is enabled. Defaults to false.
 	//+optional
```
Comment (mdbooth) [inline api/v1beta1/types.go:603]: Can we make this required with an explicit default of false?

File: controllers/openstackcluster_controller.go [hunk 2]
```diff
@@ -504,9 +507,9 @@ func bastionAddFloatingIP(openStackCluster *infrav1.OpenStackCluster, clusterNam
 	case openStackCluster.Status.Bastion.FloatingIP != "":
 		// Some floating IP has already been created for this bastion, make sure we re-use it
 		floatingIP = &openStackCluster.Status.Bastion.FloatingIP
-	case openStackCluster.Spec.Bastion.FloatingIP != "":
+	case openStackCluster.Spec.Bastion.FloatingIP != nil && *openStackCluster.Spec.Bastion.FloatingIP != "":
```
Comment (mdbooth) [inline controllers/openstackcluster_controller.go:510]: We *should* be able to just use a nil check here. Should...

File: controllers/openstackcluster_controller_test.go [hunk 3]
```diff
@@ -61,6 +61,11 @@ var _ = Describe("OpenStackCluster controller", func() {
 	capiClusterName := "capi-cluster"
 	testClusterName := "test-cluster"
 	testNum := 0
+	instance := infrav1.OpenStackMachineSpec{
```
Comment (mdbooth) [inline controllers/openstackcluster_controller_test.go:64]: nit
```suggestion
	bastionSpec := infrav1.OpenStackMachineSpec{
```

File: controllers/openstackcluster_controller_test.go [hunk 4]
```diff
@@ -194,6 +199,7 @@ var _ = Describe("OpenStackCluster controller", func() {
 		testCluster.Spec = infrav1.OpenStackClusterSpec{
 			Bastion: &infrav1.Bastion{
 				Enabled: false,
+				Spec:    &instance,
```
Comment (mdbooth) [inline controllers/openstackcluster_controller_test.go:202]: You can (should?) omit the Spec here.

File: controllers/openstackcluster_controller_test.go [hunk 5]
```diff
@@ -471,7 +480,10 @@ var _ = Describe("OpenStackCluster controller", func() {
 	It("should delete an existing bastion even if its uuid is not stored in status", func() {
 		testCluster.SetName("delete-existing-bastion")
 		testCluster.Spec = infrav1.OpenStackClusterSpec{
-			Bastion: &infrav1.Bastion{},
+			Bastion: &infrav1.Bastion{
+				Enabled: true,
+				Spec:    &instance,
+			},
```
Comment (mdbooth) [inline controllers/openstackcluster_controller_test.go:486]: I don't think this test is equivalent. The previous test had: `Bastion: &infrav1.Bastion{}`, which implied `Enabled: false`, and we apparently still expect a bastion to be cleaned up in this case. Is that what we were actually testing before? Does it still work that way?

File: controllers/openstackmachine_controller.go [hunk 6]
```diff
@@ -586,7 +586,7 @@ func (r *OpenStackMachineReconciler) reconcileNormal(ctx context.Context, scope
 		// The other state is normal (for example, migrating, shutoff) but we don't want to proceed until it's ACTIVE
 		// due to potential conflict or unexpected actions
 		scope.Logger().Info("Waiting for instance to become ACTIVE", "id", instanceStatus.ID(), "status", instanceStatus.State())
-		conditions.MarkUnknown(openStackMachine, infrav1.InstanceReadyCondition, infrav1.InstanceNotReadyReason, "Instance state is not handled: %s", instanceStatus.State())
+		conditions.MarkUnknown(openStackMachine, infrav1.InstanceReadyCondition, infrav1.InstanceNotReadyReason, "Spec state is not handled: %s", instanceStatus.State())
```
Comment (mdbooth) [inline controllers/openstackmachine_controller.go:589]: Over-aggressive search/replace!

```suggestion
		conditions.MarkUnknown(openStackMachine, infrav1.InstanceReadyCondition, infrav1.InstanceNotReadyReason, "Instance state is not handled: %s", instanceStatus.State())
```

File: pkg/cloud/services/compute/instance_types.go [hunk 7]
```diff
@@ -30,7 +30,7 @@ import (
 
 // InstanceSpec defines the fields which can be set on a new OpenStack instance.
 //
-// InstanceSpec does not contain all of the fields of infrav1.Instance, as not
+// InstanceSpec does not contain all of the fields of infrav1.Spec, as not
```
Comment (mdbooth) [inline pkg/cloud/services/compute/instance_types.go:33]: This doesn't look correct.

Also, this comment was relevant context when I originally forked this struct from Instance. I think we can just delete it now.

File: api/v1alpha6/types_conversion.go [hunk 8]
```diff
@@ -307,7 +307,7 @@ func Convert_v1beta1_BindingProfile_To_Map_string_To_Interface(in *infrav1.Bindi
 
 /* FixedIP */
 /* AddressPair */
-/* Instance */
+/* Spec */
```
Comment (mdbooth) [inline api/v1alpha6/types_conversion.go:310]: Nope?

File: test/e2e/suites/apivalidations/openstackcluster_test.go
Comment (mdbooth) [inline test/e2e/suites/apivalidations/openstackcluster_test.go:1]: Nice tests!

File: api/v1beta1/types.go (hunk 1 above)
Comment (dulek) [reply api/v1beta1/types.go:603]: Thanks for this, I wasn't sure how to approach this.

File: controllers/openstackcluster_controller_test.go (hunk 5 above)
Comment (dulek) [reply controllers/openstackcluster_controller_test.go:486]: Okay, I needed to actually fix this with a hack on `bastionToInstanceSpec()`, where nil `.spec` will be translated to `&OpenStackMachineSpec{}`.

**Pull request merged**